langchain4j/src/main/java/dev/langchain4j/data/document/splitter/DocumentByMarkdownHeaderSplitter.java

```diff
@@ -0,0 +1,159 @@
+package dev.langchain4j.data.document.splitter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import dev.langchain4j.data.document.DocumentSplitter;
+import dev.langchain4j.model.Tokenizer;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+
+public class DocumentByMarkdownHeaderSplitter extends HierarchicalDocumentSplitter {
```

Comment:
Could you please add some javadoc explaining how this splitter works? See `DocumentByParagraphSplitter` as an example.